dbms/src/Functions/FunctionsArithmetic.h

```diff
@@ -293,6 +397,19 @@ struct BitRotateLeftImpl
         return (static_cast<Result>(a) << static_cast<Result>(b))
             | (static_cast<Result>(a) >> ((sizeof(Result) * 8) - static_cast<Result>(b)));
     }
+
+#if USE_EMBEDDED_COMPILER
+    static constexpr bool compilable = true;
+
+    static inline llvm::Value * compile(llvm::IRBuilder<> & b, llvm::Value * left, llvm::Value * right, bool)
+    {
+        if (!left->getType()->isIntegerTy())
+            throw Exception("BitRotateLeftImpl expected an integral type", ErrorCodes::LOGICAL_ERROR);
+        auto * size = llvm::ConstantInt::get(left->getType(), left->getType()->getPrimitiveSizeInBits());
+        /// XXX how is this supposed to behave in signed mode?
```

Comment:
Any reasonable implementation specific behaviour.
PS. Doesn't LLVM contain "rotate" instructions?

Reply:
The problem here is that only `execute` is implementation-defined, while behavior of `compile` depends only on whether `lshr` or `ashr` is used. This means enabling compilation may change behavior.

>PS. Doesn't LLVM contain "rotate" instructions?

Not any I know of. The only reference I could find is a discussion about MIPS `rotr` instructions on the llvm-dev mailing list, where the bottom line is that a dedicated "rotate" instruction was never added to LLVM because it would increase the optimizer's complexity.